frame/support/src/traits.rs

```diff
@@ -1518,6 +1518,24 @@ pub trait OnFinalize<BlockNumber> {
 	fn on_finalize(_n: BlockNumber) {}
 }
 
+pub trait OnIdle<BlockNumber> {
+	/// The block is being finalized. Implement to have something happen.
+	///
+	/// NOTE: This function is called AFTER ALL extrinsics in a block are applied,
+	/// including inherent extrinsics.
+	fn on_idle(_n: BlockNumber, _remaining_weight: crate::weights::Weight) -> crate::weights::Weight { 0 }
+}
+
+#[impl_for_tuples(30)]
+impl<BlockNumber: Clone> OnIdle<BlockNumber> for Tuple {
+	fn on_idle(_n: BlockNumber,  _remaining_weight: crate::weights::Weight) -> crate::weights::Weight {
+		let mut weight = 0;
+		let adjusted_remaining_weight = if weight < _remaining_weight {_remaining_weight.saturating_sub(weight)} else {0};
+		for_tuples!( #( weight = weight.saturating_add(Tuple::on_idle(_n.clone(),  adjusted_remaining_weight.clone())); )* );
```

Comment:
You need to adjust at each step
```suggestion
		
		for_tuples!( 
		#( 
		let adjusted_remaining_weight = if weight < _remaining_weight {_remaining_weight.saturating_sub(weight)} else {0};
		weight = weight.saturating_add(Tuple::on_idle(_n.clone(),  adjusted_remaining_weight.clone())); 
		)* 
		);
```